File: websocket_service/clients/python/gateway/websocket_gateway.py
"""
WebSocket Gateway for Python VAD Client

This module provides the WebSocket gateway that manages connections to the VAD server,
handles configuration, and streams audio data while receiving VAD events.
"""

import asyncio
import json
import logging
import time
from typing import Optional, Dict, Any, Callable, List, TYPE_CHECKING
from urllib.parse import urlencode

import websockets
from websockets.exceptions import ConnectionClosed, WebSocketException

if TYPE_CHECKING:
    from sources.audio_sources import AudioSource


class VADEventHandler:
    """Handler for VAD events received from server."""
    
    def __init__(self):
        self.events: List[Dict[str, Any]] = []
        self.segment_count = 0
        self.logger = logging.getLogger(f"{self.__class__.__name__}")
    
    async def handle_event(self, event_data: Dict[str, Any]):
        """
        Handle a VAD event from the server.
        
        Args:
            event_data: Event data as dictionary
        """
        event_type = event_data.get("event")
        timestamp = event_data.get("timestamp_ms", 0)
        
        # Store event
        self.events.append(event_data)
        
        # Log event
        if event_type == "VOICE_START":
            segment_index = event_data.get("segment_index", 0)
            self.logger.info(f"🎙️  VOICE START - Segment #{segment_index} at {self._format_timestamp(timestamp)}")
            
        elif event_type == "VOICE_END":
            segment_index = event_data.get("segment_index", 0)
            duration = event_data.get("duration_ms", 0)
            self.segment_count = max(self.segment_count, segment_index + 1)
            self.logger.info(f"🔴 VOICE END - Segment #{segment_index}, Duration: {duration}ms")
            
        elif event_type == "VOICE_CONTINUE":
            segment_index = event_data.get("segment_index", 0)
            self.logger.debug(f"🔄 VOICE CONTINUE - Segment #{segment_index}")
            
        elif event_type == "TIMEOUT":
            message = event_data.get("message", "")
            self.logger.warning(f"⏰ TIMEOUT - {message}")
            
        elif event_type == "ERROR":
            message = event_data.get("message", "")
            self.logger.error(f"❌ ERROR - {message}")
            
        elif event_type == "INFO":
            message = event_data.get("message", "")
            self.logger.info(f"ℹ️  INFO - {message}")
        
        else:
            self.logger.debug(f"Unknown event type: {event_type}")
    
    def _format_timestamp(self, timestamp_ms: int) -> str:
        """Format timestamp for display."""
        timestamp_s = timestamp_ms / 1000
        return time.strftime("%H:%M:%S", time.localtime(timestamp_s))
    
    def get_voice_segments(self) -> List[Dict[str, Any]]:
        """Get list of detected voice segments."""
        segments = []
        
        # Group START/END events by segment_index
        start_events = {e["segment_index"]: e for e in self.events if e.get("event") == "VOICE_START"}
        end_events = {e["segment_index"]: e for e in self.events if e.get("event") == "VOICE_END"}
        
        for segment_index in sorted(set(start_events.keys()) | set(end_events.keys())):
            start_event = start_events.get(segment_index)
            end_event = end_events.get(segment_index)
            
            if start_event and end_event:
                segments.append({
                    "segment_index": segment_index,
                    "start_ms": start_event["timestamp_ms"],
                    "end_ms": end_event["timestamp_ms"],
                    "duration_ms": end_event.get("duration_ms", 0)
                })
        
        return segments
    
    def get_statistics(self) -> Dict[str, Any]:
        """Get processing statistics."""
        segments = self.get_voice_segments()
        
        return {
            "total_events": len(self.events),
            "voice_segments": len(segments),
            "total_voice_duration_ms": sum(s["duration_ms"] for s in segments),
            "events_by_type": {
                event_type: len([e for e in self.events if e.get("event") == event_type])
                for event_type in ["VOICE_START", "VOICE_END", "VOICE_CONTINUE", "TIMEOUT", "ERROR", "INFO"]
            }
        }


class WebSocketGateway:
    """WebSocket gateway for VAD server communication."""
    
    def __init__(self, 
                 server_url: str = "ws://localhost:8000/vad",
                 config: Optional[Dict[str, Any]] = None,
                 event_handler: Optional[VADEventHandler] = None):
        """
        Initialize WebSocket gateway.
        
        Args:
            server_url: WebSocket server URL (without query params)
            config: Initial configuration parameters
            event_handler: Event handler for VAD events
        """
        self.server_url = server_url
        self.config = config or {}
        self.event_handler = event_handler or VADEventHandler()
        
        self.websocket: Optional[websockets.WebSocketServerProtocol] = None
        self.connected = False
        self.reconnect_attempts = 0
        self.max_reconnect_attempts = 5
        
        self.logger = logging.getLogger(f"{self.__class__.__name__}")
    
    def _build_connection_url(self) -> str:
        """Build WebSocket connection URL with query parameters."""
        if not self.config:
            return self.server_url
        
        # Build query string from config
        query_params = {}
        for key, value in self.config.items():
            if value is not None:
                query_params[key] = str(value)
        
        if query_params:
            query_string = urlencode(query_params)
            return f"{self.server_url}?{query_string}"
        else:
            return self.server_url
    
    async def connect(self) -> bool:
        """
        Connect to WebSocket server.
        
        Returns:
            True if connection successful, False otherwise
        """
        try:
            connection_url = self._build_connection_url()
            self.logger.info(f"Connecting to {connection_url}")
            
            self.websocket = await websockets.connect(
                connection_url,
                ping_interval=30,
                ping_timeout=10,
                close_timeout=10
            )
            
            self.connected = True
            self.reconnect_attempts = 0
            self.logger.info("WebSocket connection established")
            
            return True
            
        except Exception as e:
            self.logger.error(f"Failed to connect: {e}")
            self.connected = False
            return False
    
    async def disconnect(self):
        """Disconnect from WebSocket server."""
        if self.websocket:
            await self.websocket.close()
            self.websocket = None
        
        self.connected = False
        self.logger.info("WebSocket connection closed")
    
    async def send_config(self, config_update: Dict[str, Any]):
        """
        Send configuration update to server.
        
        Args:
            config_update: Configuration parameters to update
        """
        if not self.connected or not self.websocket:
            raise RuntimeError("Not connected to server")
        
        config_message = {
            "type": "CONFIG",
            **config_update
        }
        
        try:
            await self.websocket.send(json.dumps(config_message))
            self.logger.info(f"Sent configuration update: {config_update}")
            
        except Exception as e:
            self.logger.error(f"Failed to send config: {e}")
            raise
    
    async def send_audio_frame(self, frame_data: bytes):
        """
        Send audio frame to server.
        
        Args:
            frame_data: PCM audio frame as bytes
        """
        if not self.connected or not self.websocket:
            raise RuntimeError("Not connected to server")
        
        try:
            await self.websocket.send(frame_data)
            
        except Exception as e:
            self.logger.error(f"Failed to send audio frame: {e}")
            raise
    
    async def send_heartbeat(self):
        """Send heartbeat message to server."""
        if not self.connected or not self.websocket:
            return
        
        heartbeat_message = {"type": "HEARTBEAT"}
        
        try:
            await self.websocket.send(json.dumps(heartbeat_message))
            self.logger.debug("Sent heartbeat")
            
        except Exception as e:
            self.logger.error(f"Failed to send heartbeat: {e}")
    
    async def listen_for_events(self):
        """
        Listen for events from server in a loop.
        Should be run as a background task.
        """
        if not self.connected or not self.websocket:
            raise RuntimeError("Not connected to server")
        
        try:
            async for message in self.websocket:
                try:
                    # Parse JSON message
                    event_data = json.loads(message)
                    
                    # Handle event
                    await self.event_handler.handle_event(event_data)
                    
                except json.JSONDecodeError as e:
                    self.logger.error(f"Failed to parse server message: {e}")
                except Exception as e:
                    self.logger.error(f"Error handling event: {e}")
        
        except ConnectionClosed:
            self.logger.warning("Server connection closed")
            self.connected = False
        except WebSocketException as e:
            self.logger.error(f"WebSocket error: {e}")
            self.connected = False
        except Exception as e:
            self.logger.error(f"Unexpected error in event listener: {e}")
            self.connected = False
    
    async def stream_audio(self, audio_source: "AudioSource", duration_seconds: Optional[float] = None):
        """
        Stream audio from source to server.
        
        Args:
            audio_source: Audio source to stream from
            duration_seconds: Maximum duration to stream (None for unlimited)
        """
        if not self.connected or not self.websocket:
            raise RuntimeError("Not connected to server")
        
        self.logger.info("Starting audio stream")
        
        start_time = time.time()
        frame_count = 0
        
        try:
            async for frame_data in audio_source.generate_frames():
                # Check duration limit
                if duration_seconds and (time.time() - start_time) >= duration_seconds:
                    self.logger.info(f"Reached duration limit: {duration_seconds}s")
                    break
                
                # Send frame
                await self.send_audio_frame(frame_data)
                frame_count += 1
                
                # Log progress occasionally
                if frame_count % 100 == 0:
                    elapsed = time.time() - start_time
                    self.logger.debug(f"Streamed {frame_count} frames in {elapsed:.1f}s")
        
        except Exception as e:
            self.logger.error(f"Error during audio streaming: {e}")
            raise
        
        finally:
            elapsed = time.time() - start_time
            self.logger.info(f"Audio streaming completed: {frame_count} frames in {elapsed:.1f}s")
    
    async def run_with_reconnect(self, 
                                audio_source: "AudioSource", 
                                duration_seconds: Optional[float] = None):
        """
        Run audio streaming with automatic reconnection.
        
        Args:
            audio_source: Audio source to stream from
            duration_seconds: Maximum duration to stream
        """
        while self.reconnect_attempts < self.max_reconnect_attempts:
            try:
                # Connect if not connected
                if not self.connected:
                    if not await self.connect():
                        self.reconnect_attempts += 1
                        await asyncio.sleep(2 ** self.reconnect_attempts)  # Exponential backoff
                        continue
                
                # Start event listener
                event_task = asyncio.create_task(self.listen_for_events())
                
                # Start audio streaming
                stream_task = asyncio.create_task(
                    self.stream_audio(audio_source, duration_seconds)
                )
                
                # Wait for either task to complete
                done, pending = await asyncio.wait(
                    [event_task, stream_task],
                    return_when=asyncio.FIRST_COMPLETED
                )
                
                # Cancel remaining tasks
                for task in pending:
                    task.cancel()
                    try:
                        await task
                    except asyncio.CancelledError:
                        pass
                
                # Check for exceptions
                for task in done:
                    if task.exception():
                        raise task.exception()
                
                # If we get here, streaming completed successfully
                break
                
            except (ConnectionClosed, WebSocketException) as e:
                self.logger.warning(f"Connection lost: {e}")
                self.connected = False
                self.reconnect_attempts += 1
                
                if self.reconnect_attempts < self.max_reconnect_attempts:
                    backoff_time = 2 ** self.reconnect_attempts
                    self.logger.info(f"Reconnecting in {backoff_time}s (attempt {self.reconnect_attempts})")
                    await asyncio.sleep(backoff_time)
                else:
                    self.logger.error("Max reconnection attempts reached")
                    raise
            
            except Exception as e:
                self.logger.error(f"Unexpected error: {e}")
                raise
        
        if self.reconnect_attempts >= self.max_reconnect_attempts:
            raise RuntimeError("Failed to establish stable connection after maximum attempts")


# Example usage and testing
async def test_websocket_gateway():
    """Test function for WebSocketGateway."""
    logging.basicConfig(level=logging.INFO)
    
    # Test configuration
    config = {
        "mode": "pcm",
        "sample_rate": 16000,
        "channels": 1,
        "sample_width": 2,
        "frame_duration_ms": 30,
        "start_probability": 0.4,
        "end_probability": 0.3,
        "start_frame_count": 6,
        "end_frame_count": 12,
        "timeout": 30.0
    }
    
    # Create gateway
    gateway = WebSocketGateway(
        server_url="ws://localhost:8000/vad",
        config=config
    )
    
    try:
        # Connect
        if await gateway.connect():
            print("Connected successfully")
            
            # Send heartbeat
            await gateway.send_heartbeat()
            
            # Listen for a few seconds
            event_task = asyncio.create_task(gateway.listen_for_events())
            await asyncio.sleep(5)
            event_task.cancel()
            
            # Print statistics
            stats = gateway.event_handler.get_statistics()
            print(f"Statistics: {stats}")
            
        else:
            print("Failed to connect")
    
    except Exception as e:
        print(f"Test failed: {e}")
    
    finally:
        await gateway.disconnect()


if __name__ == "__main__":
    asyncio.run(test_websocket_gateway())

File: websocket_service/clients/java/gateway/WebSocketGateway.java
/**
 * WebSocket Gateway for Java VAD Client
 * 
 * Manages WebSocket connections to the VAD server using only standard Java libraries.
 * Note: This implementation uses basic WebSocket support without external libraries.
 */

import java.io.*;
import java.net.*;
import java.nio.ByteBuffer;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.util.*;
import java.util.concurrent.*;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.logging.Logger;
import java.util.logging.Level;

public class WebSocketGateway {
    private static final Logger logger = Logger.getLogger(WebSocketGateway.class.getName());
    
    private static final String WEBSOCKET_MAGIC_STRING = "258EAFA5-E914-47DA-95CA-C5AB0DC85B11";
    private static final int FRAME_OPCODE_TEXT = 0x1;
    private static final int FRAME_OPCODE_BINARY = 0x2;
    private static final int FRAME_OPCODE_CLOSE = 0x8;
    private static final int FRAME_OPCODE_PING = 0x9;
    private static final int FRAME_OPCODE_PONG = 0xA;
    
    private String serverUrl;
    private Map<String, String> config;
    private VADEventHandler eventHandler;
    
    private Socket socket;
    private InputStream inputStream;
    private OutputStream outputStream;
    private AtomicBoolean connected;
    private AtomicBoolean running;
    private ExecutorService executorService;
    
    public WebSocketGateway(String serverUrl, Map<String, String> config, VADEventHandler eventHandler) {
        this.serverUrl = serverUrl;
        this.config = config != null ? new HashMap<>(config) : new HashMap<>();
        this.eventHandler = eventHandler != null ? eventHandler : new VADEventHandler();
        
        this.connected = new AtomicBoolean(false);
        this.running = new AtomicBoolean(false);
        this.executorService = Executors.newCachedThreadPool();
    }
    
    /**
     * Build WebSocket connection URL with query parameters.
     */
    private String buildConnectionUrl() {
        if (config.isEmpty()) {
            return serverUrl;
        }
        
        StringBuilder queryString = new StringBuilder();
        for (Map.Entry<String, String> entry : config.entrySet()) {
            if (queryString.length() > 0) {
                queryString.append("&");
            }
            try {
                queryString.append(URLEncoder.encode(entry.getKey(), "UTF-8"))
                          .append("=")
                          .append(URLEncoder.encode(entry.getValue(), "UTF-8"));
            } catch (UnsupportedEncodingException e) {
                // UTF-8 is always supported
                throw new RuntimeException(e);
            }
        }
        
        if (queryString.length() > 0) {
            return serverUrl + "?" + queryString.toString();
        } else {
            return serverUrl;
        }
    }
    
    /**
     * Connect to WebSocket server.
     */
    public boolean connect() throws Exception {
        try {
            String connectionUrl = buildConnectionUrl();
            logger.info("Connecting to " + connectionUrl);
            
            // Parse URL
            URI uri = new URI(connectionUrl);
            String host = uri.getHost();
            int port = uri.getPort();
            if (port == -1) {
                port = "wss".equals(uri.getScheme()) ? 443 : 80;
            }
            String path = uri.getPath();
            if (uri.getQuery() != null) {
                path += "?" + uri.getQuery();
            }
            
            // Create socket connection
            socket = new Socket(host, port);
            inputStream = socket.getInputStream();
            outputStream = socket.getOutputStream();
            
            // Perform WebSocket handshake
            if (performHandshake(host, path)) {
                connected.set(true);
                running.set(true);
                
                // Start message listener
                executorService.submit(this::messageListener);
                
                logger.info("WebSocket connection established");
                return true;
            } else {
                disconnect();
                return false;
            }
            
        } catch (Exception e) {
            logger.log(Level.SEVERE, "Failed to connect", e);
            connected.set(false);
            return false;
        }
    }
    
    /**
     * Perform WebSocket handshake.
     */
    private boolean performHandshake(String host, String path) throws Exception {
        // Generate WebSocket key
        byte[] keyBytes = new byte[16];
        new Random().nextBytes(keyBytes);
        String webSocketKey = Base64.getEncoder().encodeToString(keyBytes);
        
        // Build handshake request
        StringBuilder request = new StringBuilder();
        request.append("GET ").append(path).append(" HTTP/1.1\r\n");
        request.append("Host: ").append(host).append("\r\n");
        request.append("Upgrade: websocket\r\n");
        request.append("Connection: Upgrade\r\n");
        request.append("Sec-WebSocket-Key: ").append(webSocketKey).append("\r\n");
        request.append("Sec-WebSocket-Version: 13\r\n");
        request.append("\r\n");
        
        // Send handshake request
        outputStream.write(request.toString().getBytes(StandardCharsets.UTF_8));
        outputStream.flush();
        
        // Read handshake response
        BufferedReader reader = new BufferedReader(new InputStreamReader(inputStream, StandardCharsets.UTF_8));
        String statusLine = reader.readLine();
        
        if (statusLine == null || !statusLine.contains("101")) {
            logger.severe("WebSocket handshake failed: " + statusLine);
            return false;
        }
        
        // Read headers
        Map<String, String> headers = new HashMap<>();
        String line;
        while ((line = reader.readLine()) != null && !line.isEmpty()) {
            int colonIndex = line.indexOf(':');
            if (colonIndex > 0) {
                String headerName = line.substring(0, colonIndex).trim().toLowerCase();
                String headerValue = line.substring(colonIndex + 1).trim();
                headers.put(headerName, headerValue);
            }
        }
        
        // Verify handshake response
        String expectedAccept = calculateWebSocketAccept(webSocketKey);
        String actualAccept = headers.get("sec-websocket-accept");
        
        if (!expectedAccept.equals(actualAccept)) {
            logger.severe("WebSocket handshake verification failed");
            return false;
        }
        
        logger.info("WebSocket handshake successful");
        return true;
    }
    
    /**
     * Calculate WebSocket accept value.
     */
    private String calculateWebSocketAccept(String webSocketKey) throws Exception {
        String combined = webSocketKey + WEBSOCKET_MAGIC_STRING;
        MessageDigest digest = MessageDigest.getInstance("SHA-1");
        byte[] hash = digest.digest(combined.getBytes(StandardCharsets.UTF_8));
        return Base64.getEncoder().encodeToString(hash);
    }
    
    /**
     * Disconnect from WebSocket server.
     */
    public void disconnect() {
        running.set(false);
        connected.set(false);
        
        try {
            if (socket != null && !socket.isClosed()) {
                socket.close();
            }
        } catch (IOException e) {
            logger.log(Level.WARNING, "Error closing socket", e);
        }
        
        logger.info("WebSocket connection closed");
    }
    
    /**
     * Send configuration update to server.
     */
    public void sendConfig(Map<String, Object> configUpdate) throws Exception {
        if (!connected.get()) {
            throw new RuntimeException("Not connected to server");
        }
        
        Map<String, Object> configMessage = new HashMap<>();
        configMessage.put("type", "CONFIG");
        configMessage.putAll(configUpdate);
        
        String json = mapToJson(configMessage);
        sendTextFrame(json);
        
        logger.info("Sent configuration update: " + configUpdate);
    }
    
    /**
     * Send audio frame to server.
     */
    public void sendAudioFrame(byte[] frameData) throws Exception {
        if (!connected.get()) {
            throw new RuntimeException("Not connected to server");
        }
        
        sendBinaryFrame(frameData);
    }
    
    /**
     * Send heartbeat message to server.
     */
    public void sendHeartbeat() throws Exception {
        if (!connected.get()) {
            return;
        }
        
        Map<String, Object> heartbeatMessage = new HashMap<>();
        heartbeatMessage.put("type", "HEARTBEAT");
        
        String json = mapToJson(heartbeatMessage);
        sendTextFrame(json);
        
        logger.fine("Sent heartbeat");
    }
    
    /**
     * Send text frame.
     */
    private void sendTextFrame(String text) throws Exception {
        byte[] payload = text.getBytes(StandardCharsets.UTF_8);
        sendFrame(FRAME_OPCODE_TEXT, payload);
    }
    
    /**
     * Send binary frame.
     */
    private void sendBinaryFrame(byte[] data) throws Exception {
        sendFrame(FRAME_OPCODE_BINARY, data);
    }
    
    /**
     * Send WebSocket frame.
     */
    private synchronized void sendFrame(int opcode, byte[] payload) throws Exception {
        if (outputStream == null) {
            throw new IOException("Output stream not available");
        }
        
        // Create frame header
        ByteArrayOutputStream frame = new ByteArrayOutputStream();
        
        // First byte: FIN = 1, RSV = 0, Opcode
        frame.write(0x80 | opcode);
        
        // Payload length and mask
        boolean masked = true; // Client frames should be masked
        int payloadLength = payload.length;
        
        if (payloadLength < 126) {
            frame.write((masked ? 0x80 : 0) | payloadLength);
        } else if (payloadLength < 65536) {
            frame.write((masked ? 0x80 : 0) | 126);
            frame.write((payloadLength >> 8) & 0xFF);
            frame.write(payloadLength & 0xFF);
        } else {
            frame.write((masked ? 0x80 : 0) | 127);
            frame.write(0); // Extended payload length (high 32 bits)
            frame.write(0);
            frame.write(0);
            frame.write(0);
            frame.write((payloadLength >> 24) & 0xFF); // Extended payload length (low 32 bits)
            frame.write((payloadLength >> 16) & 0xFF);
            frame.write((payloadLength >> 8) & 0xFF);
            frame.write(payloadLength & 0xFF);
        }
        
        // Masking key and masked payload
        if (masked) {
            byte[] maskingKey = new byte[4];
            new Random().nextBytes(maskingKey);
            frame.write(maskingKey);
            
            // Mask payload
            for (int i = 0; i < payload.length; i++) {
                frame.write(payload[i] ^ maskingKey[i % 4]);
            }
        } else {
            frame.write(payload);
        }
        
        // Send frame
        outputStream.write(frame.toByteArray());
        outputStream.flush();
    }
    
    /**
     * Message listener thread.
     */
    private void messageListener() {
        try {
            while (running.get() && connected.get()) {
                try {
                    // Read frame
                    WebSocketFrame frame = readFrame();
                    if (frame != null) {
                        handleFrame(frame);
                    }
                } catch (SocketTimeoutException e) {
                    // Timeout is normal, continue
                } catch (IOException e) {
                    if (running.get()) {
                        logger.log(Level.WARNING, "Connection lost", e);
                        connected.set(false);
                    }
                    break;
                }
            }
        } catch (Exception e) {
            logger.log(Level.SEVERE, "Error in message listener", e);
            connected.set(false);
        }
    }
    
    /**
     * Read WebSocket frame.
     */
    private WebSocketFrame readFrame() throws IOException {
        if (inputStream.available() < 2) {
            Thread.yield();
            return null;
        }
        
        // Read first two bytes
        int byte1 = inputStream.read();
        int byte2 = inputStream.read();
        
        if (byte1 == -1 || byte2 == -1) {
            throw new IOException("Unexpected end of stream");
        }
        
        boolean fin = (byte1 & 0x80) != 0;
        int opcode = byte1 & 0x0F;
        boolean masked = (byte2 & 0x80) != 0;
        int payloadLength = byte2 & 0x7F;
        
        // Extended payload length
        if (payloadLength == 126) {
            int len1 = inputStream.read();
            int len2 = inputStream.read();
            if (len1 == -1 || len2 == -1) {
                throw new IOException("Unexpected end of stream");
            }
            payloadLength = (len1 << 8) | len2;
        } else if (payloadLength == 127) {
            // Read 8 bytes for extended length (we only support up to 32-bit length)
            for (int i = 0; i < 4; i++) {
                if (inputStream.read() == -1) {
                    throw new IOException("Unexpected end of stream");
                }
            }
            
            int len1 = inputStream.read();
            int len2 = inputStream.read();
            int len3 = inputStream.read();
            int len4 = inputStream.read();
            if (len1 == -1 || len2 == -1 || len3 == -1 || len4 == -1) {
                throw new IOException("Unexpected end of stream");
            }
            payloadLength = (len1 << 24) | (len2 << 16) | (len3 << 8) | len4;
        }
        
        // Masking key
        byte[] maskingKey = null;
        if (masked) {
            maskingKey = new byte[4];
            if (inputStream.read(maskingKey) != 4) {
                throw new IOException("Unexpected end of stream");
            }
        }
        
        // Payload
        byte[] payload = new byte[payloadLength];
        int totalRead = 0;
        while (totalRead < payloadLength) {
            int read = inputStream.read(payload, totalRead, payloadLength - totalRead);
            if (read == -1) {
                throw new IOException("Unexpected end of stream");
            }
            totalRead += read;
        }
        
        // Unmask payload
        if (masked && maskingKey != null) {
            for (int i = 0; i < payload.length; i++) {
                payload[i] ^= maskingKey[i % 4];
            }
        }
        
        return new WebSocketFrame(fin, opcode, payload);
    }
    
    /**
     * Handle received frame.
     */
    private void handleFrame(WebSocketFrame frame) {
        switch (frame.opcode) {
            case FRAME_OPCODE_TEXT:
                String text = new String(frame.payload, StandardCharsets.UTF_8);
                try {
                    Map<String, Object> eventData = jsonToMap(text);
                    eventHandler.handleEvent(eventData);
                } catch (Exception e) {
                    logger.log(Level.WARNING, "Failed to parse JSON message: " + text, e);
                }
                break;
                
            case FRAME_OPCODE_BINARY:
                logger.fine("Received binary frame (unexpected from server)");
                break;
                
            case FRAME_OPCODE_CLOSE:
                logger.info("Received close frame from server");
                connected.set(false);
                break;
                
            case FRAME_OPCODE_PING:
                try {
                    sendFrame(FRAME_OPCODE_PONG, frame.payload);
                } catch (Exception e) {
                    logger.log(Level.WARNING, "Failed to send pong", e);
                }
                break;
                
            case FRAME_OPCODE_PONG:
                logger.fine("Received pong frame");
                break;
                
            default:
                logger.warning("Unknown frame opcode: " + frame.opcode);
                break;
        }
    }
    
    /**
     * Simple JSON to Map conversion (basic implementation).
     */
    private Map<String, Object> jsonToMap(String json) {
        Map<String, Object> result = new HashMap<>();
        
        // Very basic JSON parsing - only supports simple key-value pairs
        json = json.trim();
        if (json.startsWith("{") && json.endsWith("}")) {
            json = json.substring(1, json.length() - 1);
            
            String[] pairs = json.split(",");
            for (String pair : pairs) {
                String[] keyValue = pair.split(":", 2);
                if (keyValue.length == 2) {
                    String key = keyValue[0].trim().replaceAll("\"", "");
                    String value = keyValue[1].trim();
                    
                    // Remove quotes from string values
                    if (value.startsWith("\"") && value.endsWith("\"")) {
                        value = value.substring(1, value.length() - 1);
                        result.put(key, value);
                    } else if ("true".equals(value) || "false".equals(value)) {
                        result.put(key, Boolean.parseBoolean(value));
                    } else {
                        try {
                            if (value.contains(".")) {
                                result.put(key, Double.parseDouble(value));
                            } else {
                                result.put(key, Long.parseLong(value));
                            }
                        } catch (NumberFormatException e) {
                            result.put(key, value);
                        }
                    }
                }
            }
        }
        
        return result;
    }
    
    /**
     * Simple Map to JSON conversion (basic implementation).
     */
    private String mapToJson(Map<String, Object> map) {
        StringBuilder json = new StringBuilder();
        json.append("{");
        
        boolean first = true;
        for (Map.Entry<String, Object> entry : map.entrySet()) {
            if (!first) {
                json.append(",");
            }
            first = false;
            
            json.append("\"").append(entry.getKey()).append("\":");
            
            Object value = entry.getValue();
            if (value instanceof String) {
                json.append("\"").append(value).append("\"");
            } else if (value instanceof Number || value instanceof Boolean) {
                json.append(value);
            } else {
                json.append("\"").append(value.toString()).append("\"");
            }
        }
        
        json.append("}");
        return json.toString();
    }
    
    public boolean isConnected() {
        return connected.get();
    }
    
    public VADEventHandler getEventHandler() {
        return eventHandler;
    }
    
    /**
     * Clean up resources.
     */
    public void cleanup() {
        disconnect();
        executorService.shutdown();
        try {
            if (!executorService.awaitTermination(5, TimeUnit.SECONDS)) {
                executorService.shutdownNow();
            }
        } catch (InterruptedException e) {
            executorService.shutdownNow();
            Thread.currentThread().interrupt();
        }
    }
    
    /**
     * WebSocket frame representation.
     */
    private static class WebSocketFrame {
        final boolean fin;
        final int opcode;
        final byte[] payload;
        
        WebSocketFrame(boolean fin, int opcode, byte[] payload) {
            this.fin = fin;
            this.opcode = opcode;
            this.payload = payload;
        }
    }
}
